### ISSUE TITLE
Named exports generation enhancement

### ISSUE THREAD

**Issue by jsnmoon** [5to6/5to6-codemod#34]
I would like to suggest a new transformation for adding named exports corresponding to the properties in a CommonJS default object export. The following example is common for a codebase transitioning from CommonJS to ES6 modules:

**module.before.js**:
```javascript
module.exports = {
  value1: 1,
  value2: 2,
};
```

**module.after.js**:
```
export default {
  value1: 1,
  value2: 2,
};
```

**code.before.js**:
```javascript
var module = require('module.before.js');
var value1 = module.value1, 
  value2 = module.value2;
```

**code.ideal.js**:
```javascript
import { value1, value2 } from 'module.after.js';
```

Unfortunately, **code.ideal.js** [will not work](https://github.com/babel/babel/issues/2212#issuecomment-131110500) since modules are not destructurable objects. However, we can add a compatibility shim to make **code.ideal.js** valid by adding named exports to **module.after.js**:

**module.fixed.js**
```javascript
const exported = {
  value1: 1,
  value2: 2,
};
export default exported;
export const {
  value1,
  value,
} = exported;
```

Basically, this transformation allows better compatibility between CommonJS and ES2015 module notations; some people are opting to use a [Babel 6 plugin](https://github.com/59naga/babel-plugin-add-module-exports) to facilitate this behavior.

The downside to this transform is that it can cause declaration collisions since it creates new `const` declaration for each exported object's properties. Also, this transform can be unintuitive without the above example; it'll require proper documentation in the README for people to grasp its usefulness.

I'll push out a PR for this in a day or two.

**Comment by gmathieu** [comment]
@jsnmoon I don't believe this scenario is reflective of a cjs -> es module transition. Updating value1 and value2 seems out of scope or import/export because it requires updates to both the file importing the export (code.js) and export itself (module.js). 

The example below is a perfectly reasonable conversion:

**module.after.js**:
```
export default {
  value1: 1,
  value2: 2,
};
```
**code.after.js**:
```js
import module from 'module.before.js';
var value1 = module.value1, 
  value2 = module.value2;
```

On the other hand, a transition from cjs to es modules, means we have to deal with `.default`. The following transformation is acceptable as well:

**module.after.js**:
```
export default {
  value1: 1,
  value2: 2,
};
```

**code.before.js**:
```javascript
var module = require('module.before.js').default;
var value1 = module.value1, 
  value2 = module.value2;
```

In both scenarios, updating how value1 and value2 are accessed seems like a different concern.

We've gone the validation approach to handle `.default`. I proposed an enhancement to [eslint-plugin-import](https://github.com/benmosher/eslint-plugin-import/pull/755) but it's been idle. Let me know if you find it useful and I can publish it separately.

Happy to talk this through a little more in case I'm missing the point.

**Comment by jsnmoon** [comment]
@gmathieu: Hey, thanks for the thoughtful reply! 😄 Let me address two points here: 

## Migrating module syntax once valid on Babel 5

For code transpiled with Babel 5, **code.ideal.js** would actually work with **module.after.js** as written (with `value1 === 1 === true` in **code.ideal.js**). Unfortunately in this scenario, **code.ideal.js** is written with an incorrect intention of destructuring the default import; named import syntax and object destructuring syntax are not the same *(1)*. By adding named exports, I'm hoping to eliminate dependence on this out-of-spec inter-op supported by Babel 5 and enabled by [babel-plugin-add-module-exports](59naga/babel-plugin-add-module-exports). 

To reiterate: the code as originally written is incorrect, but we can correct it by adding named exports onto the exporting module. Strange that it even worked in the first place...

## Leveraging static analysis for tree shaking

My rationale for encouraging this transformation is to enable module packagers to perform better static analysis for dead code elimination (ex. [Webpack 2](https://webpack.js.org/guides/tree-shaking/)). I agree that **module.after.js** is correct as-is because it conforms to the ES2015 module spec; your updated **code.before.js** would also work as expected. However, by dynamically importing `value1` and `value2` via variable assignment, we'll lose the ability to perform tree shaking even if **code.before.js** is converted to the following: 

Hypothetical conversion of your updated **code.before.js**
```javascript
import module from 'module.before.js';
var value1 = module.value1, 
  value2 = module.value2;
```

In this case, the entirety of `module` will be imported no matter which portion of it has actually been used in the code. 

Man, just writing all that made my head spin. Let me know if anything I've said is unclear and I'll do my best to clarify.

## Footnotes

1)  An obvious reason why object destructuring and named imports are not the same would be the `as` syntax for aliasing named imports (ex. `import { foo as bar } from 'blah'`)

**Comment by gmathieu** [comment]
Thanks for clarifying.

The biggest challenge with this migration is that it requires multiple parts to bring to completion:
1. enhance export default with named exports
1. convert import default to spreads
1. remove export default when applicable

So I would ask, how much of this should the script do before we, humans, have to get our hands dirty?

**Comment by gmathieu** [comment]
@jsnmoon, I tried your PR on part of our codebase. Webpack 2 and babel loader don't seem too happy about the workaround

**Generated output:**

```
export default exported;

export const {
    init,
    routeControllers,
    isSwitchingLicense,
    acceptStateChange,
    acceptIntercept,
    isExitPreview,
    _interceptStateChange,
    _interceptAnchorClick,
    _swapControllers
} = exported;
```

**Error message**

```
ERROR in ./src/apps/app/interceptor.babel.js
Module build failed: Duplicate declaration "routeControllers"

  257 | export const {
  258 |     init,
> 259 |     routeControllers,
      |     ^
  260 |     isSwitchingLicense,
  261 |     acceptStateChange,
  262 |     acceptIntercept,

BabelLoaderError: Duplicate declaration "routeControllers"

  257 | export const {
  258 |     init,
> 259 |     routeControllers,
      |     ^
  260 |     isSwitchingLicense,
  261 |     acceptStateChange,
  262 |     acceptIntercept,
```
But if you find utility in this conversion, I'm happy to move the PR forward

**Comment by jsnmoon** [comment]
Yes, what you mention is true: for this migration to complete, all three actions will be necessary (1). My thoughts on each step: 

1) This transform addresses this step; I think we're in agreement about this.
2) This step should be possible with purely with codemods. However, it would probably benefit immensely from being able to resolve the project's packages for checking. I think it'll be a challenging transform to write due to all the edge cases the transform would be expected to handle.
3) This step probably won't be possible with just codemods. Deleting default exports without being able to distinguish whether it has been imported elsewhere sounds annoying at best and dangerous at worst.

In summary, I think the first step is probably the only step we should try to automate via codemods. As your example clearly demonstrates, it's an inherently imperfect transformation since it can cause declaration collisions... However, I found it useful for porting my current codebase and I'm hoping it'll be useful to others.

Let's move the PR forward.

**Comment by jsnmoon** [comment]
Fixed by #39!
